drivers/platform/aducm3029/uart.c

```diff
@@ -0,0 +1,401 @@
+			      &errors);
+		if (ret != ADI_UART_SUCCESS)
+			goto failure;
+	} else { //CALLBACK_MODE
+		if (extra->waiting_read_callback) {
+			errors = WAITING_FOR_CALLBACK;
+			goto failure;
+		}
+		extra->waiting_read_callback++;
+		adi_uart_SubmitRxBuffer(
+			(ADI_UART_HANDLE const)extra->uart_handler,
+			(void *const)data,
+			(uint32_t const)bytes_number,
+			bytes_number > 4 ? true : false);
+	}
+
+	return SUCCESS;
+failure:
+	extra->errors |= errors;
+	return FAILURE;
+}
+
+/**
+ * @brief Calls the low level write functions depending on the mode
+ * @param desc:	Descriptor of the UART device
+ * @param data:	Buffer with the data to be written
+ * @param bytes_number:	Number of bytes to be written. Between 1 and 1024
+ * @return \ref SUCCESS in case of success, \ref FAILURE otherwise.
+ */
+int32_t uart_write(struct uart_desc *desc, const uint8_t *data,
+		   uint32_t bytes_number)
+{
+	uint32_t errors;
+	struct aducm_uart_desc *extra;
+	extra = desc->extra;
+
+	if (bytes_number == 0 || bytes_number > MAX_BYTES) {
+		errors = BAD_INPUT_PARAMETERS;
+		goto failure;
+	}
+
+	if (extra->mode == BLOCKING_MODE) {
+		ADI_UART_RESULT ret;
+
+		ret = adi_uart_Write(
+			      (ADI_UART_HANDLE const)extra->uart_handler,
+			      (void *const)data,
+			      (uint32_t const)bytes_number,
+			      bytes_number > 4 ? true : false,
+			      &errors);
+		if (ret != ADI_UART_SUCCESS)
+			goto failure;
+	} else { //CALLBACK_MODE
+		if (extra->waiting_write_callback) {
+			errors = WAITING_FOR_CALLBACK;
+			goto failure;
+		}
+		extra->waiting_write_callback++;
+		adi_uart_SubmitTxBuffer(
+			(ADI_UART_HANDLE const)extra->uart_handler,
+			(void *const)data,
+			(uint32_t const)bytes_number,
+			bytes_number > 4 ? true : false);
+	}
+
+	return SUCCESS;
+failure:
+	extra->errors |= errors;
+	return FAILURE;
+}
+
+/**
+ * @brief Initialize the UART communication peripheral.
+ * To configure the UART, the user must set the extra parameter from param with
+ * a pointer to the configured platform specific structure \ref
+ * aducm_uart_init_param.
+ * It enables UART in bidirectional communication.
+ * The \ref CALLBACK_MODE it is not recommended to use, except for "even-driven"
+ * peripherals. For more information refer to ADuCM302x DFP Device Driver Users
+ * Guide
+ * Only one device can be initialized, device_id = 0 which correspond with the
+ * one connected to the UART0
+ * @param desc:  Descriptor of the UART device used in the call of the drivers
+ * functions
+ * @param param: Descriptor used to configure the UART device
+ * @return \ref SUCCESS in case of success, \ref FAILURE otherwise.
+ */
+int32_t uart_init(struct uart_desc **desc, struct uart_init_param *param)
+{
+	ADI_UART_RESULT			uart_ret;
+	struct aducm_uart_desc		*aducm_desc;
+	struct aducm_uart_init_param	*aducm_init_param;
+
+	if (!desc || !param || !(param->extra) ||
+	    param->device_id != 0 || //Only UART0 available
+	    initialized != 0) //Already initialized
```

Comment:
Probably "initialized" global variable and "initialized != 0" check should be removed, since 'uart_init" should support multiple UART instances. "adi_uart_Open" also seems to support multiple UART instances, trough "param->device_id".
An opinion from another colleague is needed here.

Reply:
Info: the adcum3029 only support UART0 and the ADI drivers used are for ADuCM302x (ADuCM3029 and ADuCM3027) which also support only UART0.
But, in the drivers code I see if it is defined __ADUCM4x50__ then UART1 it is also available. I chose to use only UART0 it is needed to support also UART1?
@analogdevicesinc/noos opinion is needed here :)

Reply:
yes it should support both uarts since we want it to be platform agnostic and not intended just for ADCUM3029. Global variables shouldn't be used in this situation.